executor/merge_join.go

```diff
@@ -109,51 +117,75 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
 	return err
 }
 
-func (t *mergeJoinInnerTable) rowsWithSameKey() ([]chunk.Row, error) {
-	lastResultIdx := len(t.resultQueue) - 1
-	t.resourceQueue = append(t.resourceQueue, t.resultQueue[0:lastResultIdx]...)
-	t.resultQueue = t.resultQueue[lastResultIdx:]
+func (t *mergeJoinInnerTable) selectRow(row chunk.Row) {
+	t.curSel = append(t.curSel, row.Idx())
+}
+
+func (t *mergeJoinInnerTable) selectedRows() *chunk.RowContainer {
+	if t.curSel != nil {
+		t.curChk.SetSel(t.curSel)
+		t.sameKeyRows.SetTemporary(t.curChk)
+		t.curSel = nil
+	}
+	return t.sameKeyRows
+}
+
+func (t *mergeJoinInnerTable) rowsWithSameKey() (*chunk.RowContainer, error) {
+	// t.curSel sets to nil, so that it won't overwrite Sel in chunks in RowContainer,
+	// it might be unnecessary since merge join only runs single thread. However since we want to hand
+	// over the management of a chunk to the RowContainer, to keep the semantic consistent, we set it
+	// to nil.
+	t.curSel = nil
+	t.curChk.SetSel(nil)
+	err := t.sameKeyRows.Reset()
+	if err != nil {
+		return nil, err
+	}
+
 	// no more data.
 	if t.firstRow4Key == t.curIter.End() {
-		return nil, nil
+		return t.sameKeyRows, nil
 	}
-	t.sameKeyRows = t.sameKeyRows[:0]
-	t.sameKeyRows = append(t.sameKeyRows, t.firstRow4Key)
+	t.selectRow(t.firstRow4Key)
 	for {
 		selectedRow, err := t.nextRow()
 		// error happens or no more data.
 		if err != nil || selectedRow == t.curIter.End() {
 			t.firstRow4Key = t.curIter.End()
-			return t.sameKeyRows, err
+			return t.selectedRows(), err
 		}
 		compareResult := compareChunkRow(t.keyCmpFuncs, selectedRow, t.firstRow4Key, t.joinKeys, t.joinKeys)
 		if compareResult == 0 {
-			t.sameKeyRows = append(t.sameKeyRows, selectedRow)
+			t.selectRow(selectedRow)
```

Comment:
How about not using the `Chunk.SetSel([]int)` method and just appending `RowPtrs` into the `RowContainer`, then the `tmp` Chunk in it can be removed, which may make it simpler?

Reply:
Yes, that's how I implemented it before, and found severe performance regression, therefore I have to reduce the copy (runtime.memmove) by such an ugly implementation. I'll refactor this part and remove tmp using io.multireader, but AppendRow can't be used.